File: lib/formats/arc/sar_archive.rb
require 'lib/virtual_file'

# SAR archive
# Engine: Nscripter
# Extension: .sar
# Known games:
# - Tsukihime
module SarArchive
  module_function

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  class Unpacker
    def unpack(arc_file, output_files, _options)
      table = read_table(arc_file)
      read_contents(arc_file, table, output_files)
    end

    def read_table(arc_file)
      num_files,
      offset_to_files = arc_file.read(6).unpack('S>L>')
      if offset_to_files > arc_file.size
        fail RecognitionError, 'Bad offset to files'
      end

      table = []
      num_files.times do
        e = {}

        e[:name] = arc_file.read_until_zero
        e[:origin],
        e[:size] = arc_file.read(8).unpack('L>L>')

        e[:origin] += offset_to_files
        table.push(e)

        if e[:origin] + e[:size] > arc_file.size
          fail RecognitionError, 'Bad offset to file'
        end
      end
      table
    end

    def read_contents(arc_file, table, output_files)
      table.each do |e|
        output_files.write do
          data = arc_file.peek(e[:origin]) { arc_file.read(e[:size]) }
          VirtualFile.new(e[:name], data)
        end
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, _options)
      table_size = input_files.names.map { |n| n.length + 9 }.reduce(0, :+)
      offset_to_files = 6 + table_size
      arc_file.write([input_files.length, offset_to_files].pack('S>L>'))
      arc_file.write("\x00" * table_size)

      cur_data_origin = 0
      table_entries = []
      input_files.reverse_each do |file|
        data_size = file.data.length
        arc_file.write(file.data)
        table_entries.push([file.name, cur_data_origin, data_size])
        cur_data_origin += data_size
      end

      arc_file.seek(6)
      table_entries.each do |file_name, data_origin, data_size|
        arc_file.write(file_name.gsub('/', '\\'))
        arc_file.write("\0")
        arc_file.write([data_origin, data_size].pack('L>L>'))
      end
    end
  end
end

File: src/formats/arc/sar_archive.cc
// SAR archive
//
// Company:   -
// Engine:    Nscripter
// Extension: .sar
//
// Known games:
// - Tsukihime

#include "formats/arc/sar_archive.h"

namespace
{
    typedef struct
    {
        std::string name;
        uint32_t offset;
        uint32_t size;
    } TableEntry;

    std::unique_ptr<File> read_file(IO &arc_io, const TableEntry &table_entry)
    {
        std::unique_ptr<File> file(new File);
        file->name = table_entry.name;

        arc_io.seek(table_entry.offset);
        file->io.write_from_io(arc_io, table_entry.size);

        return file;
    }
}

void SarArchive::unpack_internal(File &file, FileSaver &file_saver) const
{
    uint16_t file_count = file.io.read_u16_be();
    uint32_t offset_to_files = file.io.read_u32_be();
    if (offset_to_files > file.io.size())
        throw std::runtime_error("Bad offset to files");

    std::vector<std::unique_ptr<TableEntry>> table;
    table.reserve(file_count);
    for (size_t i = 0; i < file_count; i ++)
    {
        std::unique_ptr<TableEntry> entry(new TableEntry);
        entry->name = file.io.read_until_zero();
        entry->offset = file.io.read_u32_be() + offset_to_files;
        entry->size = file.io.read_u32_be();
        if (entry->offset + entry->size > file.io.size())
            throw std::runtime_error("Bad offset to file");
        table.push_back(std::move(entry));
    }

    for (auto &table_entry : table)
        file_saver.save(read_file(file.io, *table_entry));
}
